### PR TITLE
Add SDRF file parsing

### PR DESCRIPTION
**Pull request by a-solovyev12**
Added functionality for parsing metadata files and incorporating them into SCE object produced from 10X gene expression matrices.

### CODE REVIEW

File: dropletutils-read-10x-counts.R [hunk 1]
```diff
@@ -13,14 +13,47 @@ option_list = list(
...
+    c("-m", "--metadata-files"),
+    action = "store",
+    default = NA, 
+    type = 'character',
+    help = "A comma-separated list containing one or more paths to the SDRF 
+            metadata files in .txt format, e.g. with cell type annotations.
+            Note that field names and gene ids must be consistent across the files 
+            and correspond to the matrix data."
+  ),
+  make_option(
```
Comment (pinin4fjords) [inline dropletutils-read-10x-counts.R:41]: Could you provide a short option for all arguments please?

File: dropletutils-read-10x-counts.R [hunk 2]
```diff
@@ -34,19 +67,50 @@ option_list = list(
...
+
+# metadata file logic 
+if(!is.na(opt$metadata_files)){
+  metadata_files = wsc_split_string(opt$metadata_files)
+  # check cell id column is provided
+  if(is.na(opt$cell_id_column)){
+    stop("Cell id column name for SDRF files is not provided. ")
+  }
+  metadata = lapply(metadata_files, function(x) read.csv(x, sep="\t"))
+  metadata = do.call(rbind, metadata)
```
Comment (pinin4fjords) [inline dropletutils-read-10x-counts.R]: I believe rbind() assumes the same number and identity of the columns. If I were you I would find the columns common to all metadata files and use those when combining the data frames, like:

```
common_names <- Reduce(intersect, lapply(metadata, colnames))
metadata = do.call(rbind, lapply(metadata, function(x) x[,comon_names]))
```

File: dropletutils-read-10x-counts.R [hunk 3]
```diff
@@ -34,19 +67,50 @@ option_list = list(
...
+  if(!is.na(opt$metadata_columns)){
+    metadata_columns = wsc_split_string(opt$metadata_columns)
+    if(!all(metadata_columns %in% colnames(metadata))){
+      stop("Provided metadata column names do not match existing names.")
+    } 
+    # subset by given columns
+    metadata = metadata[, metadata_columns]
+  }
 
-# Print an object summary
```
Comment (pinin4fjords) [inline dropletutils-read-10x-counts.R:48]: I think the below is too complex. You already removed potential duplicates so, so you just need to check that all cells have metadata, then do a simple merge() without sort:

```
if (! all(metadata[[opt$cell_id_column]] %in% colData(single_cell_experiment$Barcode))){
  stop("Error in matching metadata to expression matrix columns.")
}else{
  colData(single_cell_experiment) <- merge(colData(single_cell_experiment), metadata, by.x = 'Barcode', by.y = opt$cell_id_column, sort = FALSE)
}
```

**Comment by pinin4fjords** [comment]
When making PRs, could you add a description? Helps people other than you and I understand what's going on!

File: dropletutils-scripts-post-install-tests.bats [hunk 4]
```diff
@@ -15,14 +15,35 @@
...
+    echo "output = ${output}"
+ 
+    [ "$status" -eq 0 ]
+    [ -f  "$test_sdrf" ]
+}
+
+
 # Create the Matrix object
 
 @test "SingleCellExperiment object creation from 10x" {
```
Comment (pinin4fjords) [inline dropletutils-scripts-post-install-tests.bats:52]: Could we have an additional similar test to this /without/ specification of any of the metadata parameters? I want to be sure that this change does not impact additional functionality.

File: dropletutils-read-10x-counts.R [hunk 5]
```diff
@@ -13,14 +13,47 @@ option_list = list(
...
-    help = "A logical scalar indicating whether the columns of the output object should be named with the cell barcodes."
+    help = "A logical scalar indicating whether the columns of the output object
+            should be named with the cell barcodes."
+  ),
+  make_option(
+    c("-m", "--metadata-files"),
+    action = "store",
+    default = NA, 
+    type = 'character',
+    help = "A comma-separated list containing one or more paths to the SDRF
```
Comment (pinin4fjords) [inline dropletutils-read-10x-counts.R]: Please remove SDRF references, it's too Atlas-specific. A simple tab-delimited file is all that's required.

File: dropletutils-read-10x-counts.R [hunk 6]
```diff
@@ -13,14 +13,47 @@ option_list = list(
...
+    help = "A comma-separated list containing one or more directory names, each
+            corresponding to a 10X sample. Each directory should contain the 'matrix.mtx',
+            'genes.tsv' and 'barcodes.tsv' files generated by CellRanger. 
+            Alternatively, strings may contain a path to a HDF5 file in the sparse
+            matrix format generated by 10X. These can be mixed with directory 
+            names when type='auto'."
   ),
   make_option(
-    c("-c", "--col-names"),
+    c("-n", "--col-names"),
```
Comment (pinin4fjords) [inline dropletutils-read-10x-counts.R]: Please revert this to what it was, and use a different short argument for your new options. The idea is to leave the interface untouched for any outside code calling this script as before.

**Review by pinin4fjords — APPROVED** [review]
Okay, looking good now, thanks for the work @a-solovyev12. Merge at your leisure.

**Comment by a-solovyev12** [comment]
@pinin4fjords Thanks for review!

**Pull request merged**